Домашнее задание к лекции Управляющие конструкции и коллекции часть 2
Преподаватель: Константин Башевой
**Домашнее задание **

Задание 1
Дана статистика рекламных каналов по объемам продаж. Напишите скрипт, который возвращает название канала с максимальным объемом.
Т. е. в данном примере скрипт должен возвращать “yandex”.
stats = {“facebook”: 55, “yandex”: 120, “vk”: 115, “google”: 99, “email”: 42, “ok”: 98}

Задание 2
Дан список описаний товаров descriptions:
descriptions = [
[“браслет”, “xiaomi”, “mi”, “band”, “2.”, “умный”, “браслет”, “xiaomi”, “mi”, “лидер”, “в”, “линейке”, “xiaomi.”, “фитнес”, “браслет”, “для”, “android”, “4.4,”, “ios”, “7,”, “экран”, “oled.”, “лучший”, “выбор”, “xiaomi”],
[“браслет”, “huawei”, “honor”, “band”, “3.”, “фитнес”, “браслет”, “влагозащищенный”, “сенсорный”, “экран”, “android,”, “ios”, “мониторинг”, “сна”],
[“браслет”, “samsung”, “gear”, “fit2”, “pro.”, “влагозащищенный”, “сенсорный”, “amoled-экран”, “мониторинг”, “сна,”, “калорий,”, “новинка”, “samsung”],
]

Есть подозрение, что некоторые описания содержат много дубликатов одних и тех же слов. Т. е. могут быть переоптимизированными. Для каждого описания товара выведите на экран разницу между количеством уникальных слов и общим количеством слов в этом описании.

Задание 3
В примере поиска с помощью кортежей мы использовали 3 столбца. Напишите функцию, которая формирует словарь для поиска по n столбцам.

In [26]:
# Задание 1
# Дана статистика рекламных каналов по объемам продаж. 
# Напишите скрипт, который возвращает название канала с максимальным объемом.
# Т.е. в данном примере скрипт должен возвращать “yandex”.
stats = {'facebook': 55, 'yandex': 120, 'vk': 115, 'google': 99, 'email': 42, 'ok': 98}
stats

{'facebook': 55, 'yandex': 120, 'vk': 115, 'google': 99, 'email': 42, 'ok': 98}

In [24]:
# РЕШЕНИЕ Задания 1
# Используем промежуточный список с обратными ключевыми значениями {}
inverse = [(value, key) for key, value in stats.items()]
print(max(inverse)[1])

yandex


In [73]:
# Задание 2 Дан список описаний товаров descriptions: 
descriptions = [ 
  ['браслет', 'xiaomi', 'mi', 'band', '2.', 'умный', 'браслет', 'xiaomi', 'mi', 'лидер', 'в', 'линейке', 
   'xiaomi.', 'фитнес', 'браслет', 'для', 'android', '4.4,' , 'ios', '7,', 'экран', 'oled.', 'лучший', 
   'выбор', 'xiaomi'], 
    ['браслет', 'huawei', 'honor', 'band', '3.', 'фитнес', 'браслет', 'влагозащищенный', 
     'сенсорный', 'экран', 'android,', 'ios', 'мониторинг', 'сна'], 
    ['браслет', 'samsung', 'gear', 'fit2', 'pro.', 'влагозащищенный', 'сенсорный', 
     'amoled-экран', 'мониторинг', 'сна,', 'калорий,', 'новинка', 'samsung'], 
]
# Есть подозрение, что некоторые описания содержат много дубликатов одних и тех же слов. 
# Т.е. могут быть переоптимизированными. 
# Для каждого описания товара выведите на экран разницу между количеством уникальных слов и 
# общим количеством слов в этом описании.

In [75]:
#  РЕШЕНИЕ Задания 2
for i in range(3):
    print(descriptions[i][0],descriptions[i][1],':  общее количество слов=',len(descriptions[i]),
      'разница=', len(descriptions[i])-len(set(descriptions[i])))

браслет xiaomi :  общее количество слов= 25 разница= 5
браслет huawei :  общее количество слов= 14 разница= 1
браслет samsung :  общее количество слов= 13 разница= 1


In [74]:
# Задание 3 В примере поиска с помощью кортежей мы использовали 3 столбца. 
# Напишите функцию, которая формирует словарь для поиска по n столбцам.

In [131]:
#  РЕШЕНИЕ Задания 3
stats_dict={}
def find_n(n, file): # функция формирует словарь по n столбцам
    with open(file) as f:                  # открываем файл 
        for line in f:                     # построчно прочитываем 
            line=line.strip().split(',')   # элементы строки разделены запятой
            stats_dict[tuple(line[:n])]=line[n] # заполняем ключ словаря кортежем из n элементов
            
find_n(3,'stats.csv') # на входе n=3, file='stats.csv'. Словарь объявлен глобально. 

НИЖЕ ПРОВЕРКА РЕШЕНИЯ НА ИЗВЕСТНЫХ ДАННЫХ

In [132]:
line

['20', '26344', '2', '81208']

In [134]:
# stats_dict # весь словарь хотелось увидеть)))

In [104]:
# здесь семь строк словаря
i=0
for key,value in stats_dict.items():
    print(key, value)
    i=i+1
    if i not in range(7): break

('14', '26376', '3') 13301
('12', '7316', '2') 19329
('28', '2803', '2') 87381
('25', '22022', '2') 94036
('20', '13883', '5') 50722
('14', '21304', '1') 55449
('22', '2029', '2') 12482


In [105]:
line_to_find=('20', '20552', '3') 

In [148]:
%%time
for i in range(10**4):
    cost=stats_dict[line_to_find]
print(cost)

39156
CPU times: user 2.76 ms, sys: 45 µs, total: 2.8 ms
Wall time: 2.78 ms


железо - как живое???
Продолжительность счета меняется)))) одной и той же задачи)))


CPU times: user 2.81 ms, sys: 77 µs, total: 2.89 ms
Wall time: 2.88 ms
CPU times: user 2.78 ms, sys: 59 µs, total: 2.84 ms
Wall time: 2.83 ms
CPU times: user 2.84 ms, sys: 70 µs, total: 2.91 ms
Wall time: 2.9 ms
CPU times: user 3 ms, sys: 170 µs, total: 3.17 ms
Wall time: 3.09 ms 
CPU times: user 3.12 ms, sys: 118 µs, total: 3.23 ms
Wall time: 3.23 ms
CPU times: user 2.88 ms, sys: 77 µs, total: 2.96 ms
Wall time: 2.95 ms
CPU times: user 2.76 ms, sys: 45 µs, total: 2.8 ms
Wall time: 2.78 ms